### PR TITLE
Add id with ref support, solve bug(371).

### PR DESCRIPTION
**Pull request by willson-chen**

### CODE REVIEW

File: jsonschema/validators.py [hunk 1]
```diff
@@ -862,6 +868,44 @@ def resolve_remote(self, uri):
...
+
+            last_schema:
+
+                The referring upper level schema.
+
+            last_url:
+
+                Save the last URL.
+        """
+        if not isinstance(schema, dict) \
```
Comment (willson-chen) [inline jsonschema/validators.py:889]: Keep  ```"id" : "http://json-schema.org/draft-03/schema#", "id" : "http://json-schema.org/draft-04/schema#", "id" : "http://json-schema.org/draft-06/schema#", "id" : "http://json-schema.org/draft-07/schema#"```  unchanged in self.store

File: jsonschema/validators.py [hunk 2]
```diff
@@ -862,6 +868,44 @@ def resolve_remote(self, uri):
...
+                or self.resolution_scope in list_schema:
+            return
+
+        for k in schema.keys():
+            if k in [u"id", u"$id"] and isinstance(schema[k], str_types):
+
+                last_url = urljoin(last_url, schema[k], allow_fragments=True)
```
Comment (willson-chen) [inline jsonschema/validators.py:896]: Splice the value and the url of the upper level, and save it into last_url at the same time.

File: jsonschema/validators.py [hunk 3]
```diff
@@ -862,6 +868,44 @@ def resolve_remote(self, uri):
...
+                url, fragment = urldefrag(last_url)
+
+                if last_schema:
+                    self.store[url] = last_schema
```
Comment (willson-chen) [inline jsonschema/validators.py:900]: Save the subschema of the upper level into self.store in the form of `"url": { last_schema }`

File: jsonschema/validators.py [hunk 4]
```diff
@@ -862,6 +868,44 @@ def resolve_remote(self, uri):
...
+
+                if fragment:
+                    self.store[url][fragment] = schema
```
Comment (willson-chen) [inline jsonschema/validators.py:903]: Add fragment element and save it into `url": { last_schema }` which saved above, in the form of  `"url": { last_schema, "fagment": { schema } }`

File: jsonschema/validators.py [hunk 5]
```diff
@@ -862,6 +868,44 @@ def resolve_remote(self, uri):
...
+
+            backup_schema = copy.deepcopy(schema)
+            if isinstance(schema[k], dict):
+                self.store_subschema(schema[k], backup_schema, last_url)
```
Comment (willson-chen) [inline jsonschema/validators.py:907]: Recursively traverse the dictionary.

File: jsonschema/validators.py [hunk 6]
```diff
@@ -862,6 +868,44 @@ def resolve_remote(self, uri):
...
+                last_url = urljoin(last_url, schema[k], allow_fragments=True)
+                url, fragment = urldefrag(last_url)
+
+                if last_schema:
+                    self.store[url] = last_schema
+
+                if fragment:
+                    self.store[url][fragment] = schema
+
+            backup_schema = copy.deepcopy(schema)
```
Comment (willson-chen) [inline jsonschema/validators.py:905]: Deep copy prevents references.

File: jsonschema/validators.py (hunk 1 above)
Comment (willson-chen) [reply jsonschema/validators.py:889]: @Julian This if statement block is replaced with the following statement, there will be no problem with python3 and above， but python2 will report  error `exceptions.AttributeError: NoneType' object has no attribute 'find'`
```
if not isinstance(schema, dict) or schema.get(u"$schema") in meta_schemas:
     return
```
In order to maintain compatibility with python2, I keep the current if statement block

**Comment by willson-chen** [comment]
@Julian Hello, I have tested this PR could solve the following four skipped scenes，at the same time #710 #693 #341 all have been solved.
```
        or skip(
            message=bug(),
            subject="ref",
            case_description="Recursive references between schemas",
        )(test)
        or skip(
            message=bug(371),
            subject="ref",
            case_description="Location-independent identifier",
        )(test)
        or skip(
            message=bug(371),
            subject="ref",
            case_description=(
                "Location-independent identifier with absolute URI"
            ),
        )(test)
        or skip(
            message=bug(371),
            subject="ref",
            case_description=(
                "Location-independent identifier with base URI change in subschema"
            ),
        )(test)
```

**Comment by WilliamHeaven** [comment]
@Julian This PR can fix some skipped test cases. Could you please check it to see whether it meets the expectation when you have time?

File: jsonschema/validators.py [hunk 7]
```diff
@@ -862,6 +864,40 @@ def resolve_remote(self, uri):
...
+
+            schema:
+
+                The referring schema.
+
+            last_url:
+
+                The last URL.
+        """
+        if not isinstance(schema, dict):
```
Comment (Julian) [inline jsonschema/validators.py:881]: Will have a closer look, but the schema does not have to be a dict (it's an [arbitrary collections.Mapping](https://github.com/Julian/jsonschema/blob/main/jsonschema/validators.py#L933)) and does not necessarily even need to be copy-able.

File: jsonschema/validators.py [hunk 8]
```diff
@@ -862,6 +864,40 @@ def resolve_remote(self, uri):
...
+            return
+
+        for k in schema.keys():
+            if k in [u"id", u"$id"] and isinstance(schema[k], str):
```
Comment (Julian) [inline jsonschema/validators.py:885]: This should use `Validator.ID_OF` in some way.

File: jsonschema/validators.py [hunk 9]
```diff
@@ -862,6 +864,40 @@ def resolve_remote(self, uri):
             self.store[uri] = result
         return result
 
+    def store_subschema(self, schema, last_url=None):
```
Comment (Julian) [inline jsonschema/validators.py:867]: To start I'd not want this to be a public method / addition to the API.

**Comment by Julian** [comment]
Awesome! Left 3 quick comments from a first read on the approach. Will have a closer read but let me know if any of those seem unclear.

**Comment by WilliamHeaven** [comment]
@willson-chen I think this function is useful. Could you please modify it based on the review comments?

**Pull request closed without merging**